cpp/include/cudf/utilities/error.hpp

```diff
@@ -104,12 +104,15 @@ inline void throw_cuda_error(cudaError_t error, const char* file,
                   cudaGetErrorName(error) + " " + cudaGetErrorString(error)});
 }
 
-inline void check_stream(cudaStream_t stream, const char* file,
-                         unsigned int line) {
+inline void check_cuda_error(char const* file, unsigned int line,
+                             bool sync_on_stream, cudaStream_t stream=0) {
   cudaError_t error{cudaSuccess};
-  error = cudaStreamSynchronize(stream);
-  if (cudaSuccess != error) {
-    throw_cuda_error(error, file, line);
+
+  if (sync_on_stream) {
```

Comment:
I think `sync_on_stream` should be a template parameter to eliminate the runtime check of `sync_on_stream`. This operation should be as lightweight as possible in release builds.

Reply:
Also, we can simplify this a bit:
```c++
if(sync_on_stream)
   error = cudaStreamSynchronize(stream);
else
   error = cudaGetLastError();

if(cudaSuccess != error)
   throw_cuda_error(error,file,line);
```

Reply:
> I think `sync_on_stream` should be a template parameter to eliminate the runtime check of `sync_on_stream`. This operation should be as lightweight as possible in release builds.

Yeah, it will be safer even though compiler is likely to optimize this anyway; this function is inlined and sync_on_stream is constant as we call this function through macros. If compiler sees
```
if (false) {
  error = cudaStreamSynchronize(stream);
  if (cudaSuccess != error) {
    throw_cuda_error(error, file, line);
  }
}
```
compiler will remove this code (something called dead code elimination). But, this is likely not guaranteed, so I agree that the template approach might be safer.

Reply:
I agree it was likely happening anyways, but passing it as a template arg makes me feel a little bit better about it.